### PR TITLE
Add docs sections describing default and custom file names

### LINKED ISSUE #5069: Document how to change file system storage default file name
## Summary

The media pipeline, supported storage, file system storage section of the documentation begins by explaining that files are stored using SHA1 hash of their URLs.

This is a good default choice, but the documentation makes no mention of the `MediaPipeline.file_path()` method.

Presuming this is indeed the best way to get custom file names out of a file or images pipeline, this issue is to request a call out for setting a custom file name by overriding this method:

```
    def file_path(self, request, response=None, info=None, *, item=None):
        image_guid = hashlib.sha1(to_bytes(request.url)).hexdigest()
        return f'full/{image_guid}.jpg'
```

## Motivation

It appears [it used to be](https://stackoverflow.com/questions/18081997/scrapy-customize-image-pipeline-with-renaming-defualt-image-name?noredirect=1&lq=1) more tricky to customize file names.

Calling out this method may save people time.

## Describe alternatives you've considered

Leaving it as is or perhaps explaining why SHA1 was chosen over any other default file naming scheme.

### PR DESCRIPTION
**Pull request by banagale**
Resolves #5069

### CODE REVIEW

**Comment by banagale** [comment]
@Gallaecio I believe the failed check is part of #5160.

This docs addition is in response to sc[rapy/scrapy/issues/5069#issuecomment-813876104](https://github.com/scrapy/scrapy/issues/5069#issuecomment-813876104) 

Please let me know edits are needed or something is preventing a review.

File: docs/topics/media-pipeline.rst [hunk 1]
```diff
@@ -111,13 +111,14 @@ For the Images Pipeline, set the :setting:`IMAGES_STORE` setting::
 
    IMAGES_STORE = '/path/to/valid/dir'
 
-Supported Storage
+.. _topics-file-naming:
+
+File Naming
 =================
```
Comment (Gallaecio) [inline docs/topics/media-pipeline.rst:117]:
```suggestion
File Naming
===========
```

File: docs/topics/media-pipeline.rst [hunk 2]
```diff
@@ -127,9 +128,59 @@ Whose ``SHA1 hash`` is::
...
+
+   3afec3b4765f8f0a07b78f98c07b83f013567a0a.jpg
+
+Custom File Naming
+-------------------
+You may wish to use a different calculated file name for saved files.
+For example, classifying an image by including meta in the file name.
+
+Customize file names by overriding the ``file_path`` class method of your
+pipeline.
```
Comment (Gallaecio) [inline docs/topics/media-pipeline.rst:141]: “class method” may be mistaken for `@classmethod`, so just “method” would be best.

Also, “pipeline” is too generic (could be mistaken for “item pipeline”), “media pipeline” would be best.

**Comment by banagale** [comment]
Thanks for the above. I’ll have a look and update the PR.

**Comment by banagale** [comment]
Alright, I've made the suggested edits.

**Review by wRAR — APPROVED** [review]

File: docs/topics/media-pipeline.rst [hunk 3]
```diff
@@ -111,13 +111,14 @@ For the Images Pipeline, set the :setting:`IMAGES_STORE` setting::
...
 
-File system storage
--------------------
+File Naming
+===========
 
-The files are stored using a `SHA1 hash`_ of their URLs for the file names.
+Default File Naming
+-------------------
+By default, Scrapy will name files using a `SHA1 hash`_ of their origination URLs.
```
Comment (Gallaecio) [inline docs/topics/media-pipeline.rst:121]: :lipstick: 

- Keep an empty line after a header.
- Maintain the parts from the original sentence that did not require a change.
- Use [“an” before SHA1](https://www.quora.com/Should-you-use-a-or-an-before-an-acronym-that-begins-with-the-letter-S) while at it. Ideally it should also be SHA-1, but that requires changing other lines as well, so up to you.

```suggestion
Default File Naming
-------------------

By default, files are stored using an `SHA1 hash`_ of their URLs for the file names.
```

File: docs/topics/media-pipeline.rst [hunk 4]
```diff
@@ -127,9 +128,59 @@ Whose ``SHA1 hash`` is::
 
     3afec3b4765f8f0a07b78f98c07b83f013567a0a
 
-Will be downloaded and stored in the following file::
+Will be downloaded and stored in your chosen :ref:`storage method <topics-supported-storage>` using the following file name::
```
Comment (Gallaecio) [inline docs/topics/media-pipeline.rst:131]: :lipstick: 

```suggestion
Will be downloaded and stored using the following file name::
```

Or, alternatively:

```suggestion
Will be downloaded and stored using your chosen :ref:`storage method <topics-supported-storage>` and the following file name::
```

File: docs/topics/media-pipeline.rst [hunk 5]
```diff
@@ -127,9 +128,59 @@ Whose ``SHA1 hash`` is::
...
+
+Custom File Naming
+-------------------
+You may wish to use a different calculated file name for saved files.
+For example, classifying an image by including meta in the file name.
+
+Customize file names by overriding the ``file_path`` method of your
+media pipeline.
+
+For example, an image pipeline with image url::
```
Comment (Gallaecio) [inline docs/topics/media-pipeline.rst:143]:
```suggestion
For example, an image pipeline with image URL::
```

File: docs/topics/media-pipeline.rst [hunk 6]
```diff
@@ -127,9 +128,59 @@ Whose ``SHA1 hash`` is::
...
+
+   http://www.example.com/product/images/large/front/0000000004166
+
+Can be processed into a file name with a condensed hash and the perspective
+``front``::
```
Comment (Gallaecio) [inline docs/topics/media-pipeline.rst:148]:
```suggestion
Can be stored as::
```

File: docs/topics/media-pipeline.rst [hunk 7]
```diff
@@ -127,9 +128,59 @@ Whose ``SHA1 hash`` is::
...
+      image_url_hash = hashlib.shake_256(request.url.encode()).hexdigest(5)
+      image_perspective = request.url.split('/')[-2]
+      image_filename = f'{image_url_hash}_{image_perspective}.jpg'
+
+      return image_filename
+
+.. warning::
+  If your custom file name scheme relies on meta data that can vary between
+  scrapes it may lead to unexpected re-downloading of existing media using
+  new file names.
```
Comment (Gallaecio) [inline docs/topics/media-pipeline.rst:169]:
```suggestion
  If a file that has been downloaded with a given file name and your 
  ``file_path`` implementation returns a different file name for the same file,
  that file will be downloaded again with the new file name.
```

File: docs/topics/media-pipeline.rst [hunk 8]
```diff
@@ -139,6 +190,10 @@ Where:
 * ``full`` is a sub-directory to separate full images from thumbnails (if
   used). For more info see :ref:`topics-images-thumbnails`.
 
+* ``<FILE_NAME>`` is the default file name created by Scrapy or a customized file
+  name (if used).  For more info see :ref:`topics-file-naming`.
```
Comment (Gallaecio) [inline docs/topics/media-pipeline.rst:194]: :lipstick: Shorter.

```suggestion
* ``<FILE_NAME>`` is the file name assigned to the file. See :ref:`topics-file-naming`.
```

File: docs/topics/media-pipeline.rst (hunk 6 above)
Comment (banagale) [reply docs/topics/media-pipeline.rst:148]: This change removes signaling of the specific intent of the code. I chose this example because retail products so often provide multiple perspectives, i.e. front, back, left, right. And Scrapy seems to often be used for processing these images.

If this sentence is unclear or too wordy, can you suggest an alternative way to keep this "result" information?

File: docs/topics/media-pipeline.rst (hunk 7 above)
Comment (banagale) [reply docs/topics/media-pipeline.rst:169]: I tried this suggested language and found it a bit repetitive. I realize the use of "naming scheme" is not simple, so I've added an example that uses this phrase.

**Comment by banagale** [comment]
I've made changes based on this past set of recommended improvements. 

The indefinite article based on sound of an acronym was a new to me. It makes sense, and is something I'd probably do naturally, but I did not realize there was a rule.

File: docs/topics/media-pipeline.rst (hunk 6 above)
Comment (Gallaecio) [reply docs/topics/media-pipeline.rst:150]: OK. I did let myself go with the feedback :sweat_smile:

**Review by Gallaecio — APPROVED** [review]

File: docs/topics/media-pipeline.rst (hunk 6 above)
Comment (banagale) [reply docs/topics/media-pipeline.rst:150]: :) The PR is very much improved from the feedback. Thank you.

**Comment by Gallaecio** [comment]
Thanks!
